04-onion-routing.md

```diff
@@ -1407,6 +1438,88 @@ The _origin node_:
+
+Peers build reputation by forwarding successful HTLCs that resolve quickly, and
+lose reputation if they endorse failing or slow-resolving HTLCs. Reputation is
+only _negatively_ affected if an endorsed HTLC resolves undesirably, to holds
+nodes accountable for their endorsement signal while still allowing them to 
+forward unendorsed HTLCs that they are not certain about.
+
+HTLC resolution time is assessed relative to a threshold that the node 
+considers to be a reasonable amount of time for a HTLC to resolve: 
+- `resolution_period`: the amount of time a HTLC is allowed to resolve in that 
+  is classified as "good" behavior (default: 10 seconds).
+- `resolution_time`: the time elapsed between `update_add_htlc` and its 
+  resolution (`update_fulfill_hltc` / `update_fail_hltc` / 
+  `update_fail_malformed_htlc`).
+
+Each HTLC's contribution to reputation is expressed by its `effective_fees` 
+which capture the fees that it paid and the opportunity cost that its 
+`resolution_time` incurred. 
+- `fees`: the fees paid by a forwarded HTLC (as described in [BOLT #7](07-routing-gossip.md#htlc-fees), 
+  equal to 0 if the HTLC was failed).
+- `opportunity_cost`: 'ceil ( (resolution_time - resolution_period) / resolution_period) * fees`
+
+Given that `resolution_time` will be > 0 in practice, `opportunity_cost` is 0 
+for HTLCs that resolve within `resolution_period`, and charges the `fees` that 
+the HTLC would have earned per period it is held thereafter. This cost accounts 
+for the slot and liquidity that could have otherwise been paid for by 
+successful, fast resolving HTLCs during the `resolution_time` the HTLC was 
+locked in the channel.
+
+For every incoming HLTC a peer has forwarded through a node, its `effective_fees` 
+are calculated as follows:
+- if `endorsed` = 1 in `update_add_htlc`:
+  - `effective_fees` = `fees` - `opportunity_cost` 
+- otherwise: 
+  - if successfully resolved AND `resolution_time` <= `resolution_period`
+    - `effective_fees` = `fees`
+  - otherwise: 
+    - `effective_fees` = 0
+
+The `effective_fees` that a peer has paid our node are compared to our total 
+routing revenue to classify a peer's reputation as "good" or "neutral". This
+relates the fees that must be paid to earn "good" reputation to the damage that 
+can be done by abusing it.
+
+Routing revenue is assessed over a sliding window, so that reputation 
+classification is related to the node's current routing activity. Nodes MAY
+choose a window per their preferences - shorter windows being more reactive to 
+recent routing patterns, and longer windows aggregating trends over time.
+
+- `routing_window`: the period of time in which total routing revenue is 
+  assessed (default: 10 minutes * the maximum number of blocks in the future a
+  HTLC's expiry height can fall before the node will send `expiry_too_far`, as 
+  outlined in [Failure Messages](#failure-messages)).
+- `routing_revenue`: the sum of all fees paid to the node to forward HTLCs
```

Comment:
`fees paid to the node to forward HTLCs`: over a specific channel?

Reply:
No, just total revenue earned as a routing node over the window specified.

Reply:
So if I'm a big routing node with 100s of channels then in order to endorse some peer they need to have `reputation_revenue` over a big window that exceeds the total `routing_revenue` of a small window. This is more and more demanding as my channel/peer numbers go up, as it will pump `routing_revenue` whilst `reputation_revenue` stays the same.

For a node like [ACINQ](https://amboss.space/node/03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f) with ~3761 channels is it even possible that one single peer can satisfy `reputation_revenue >= routing_revenue` ?

Reply:
Why not have the `routing_revenue` be specific to the peer that is forwarding you the HTLC?
i.e (as Bob) I will endorse Alice if her `reputation_revenue` is greater than the `routing_revenue` generated by **Alice's** HTLCs.

Then `reputation_revenue >= routing_revenue` would have to be satisfied in a per-peer basis and wouldn't be influenced by the size of the node.

Reply:
The thing that we are trying to quantify here is "How much damage can Alice do?". If Bob has a channel with Carol that gives him most of the fee revenue, Alice can jam it by taking up all the slots and Bob will lose more funds than Alice has ever paid him.

Reply:
Yes Alice should not be given that much power, perhaps [this](https://github.com/lightning/bolts/pull/1071#discussion_r1197614107) can help here